omics/omics_uri_agent/uri_parser.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+from urllib.parse import urlparse
+from omics.transfer import (
+    ReadSetFileName,
+    ReferenceFileName
+)
+
+
+# Example URIs supported:
+# omics://429915189008.storage.us-east-1.amazonaws.com/1981413158/readSet/5346184667/source1
+# omics://429915189008.storage.us-east-1.amazonaws.com/1981413158/reference/5346184667/source
+class OmicsUri:
+    SCHEME = "omics"
+    NETLOC = "STORAGE"
+    IDENTIFIER = ""
+    DEFAULT_FILE_NAME = None
+
+    def __init__(self, omics_uri):
+        # https://docs.python.org/3/library/urllib.parse.html
+        url_parts = urlparse(omics_uri.upper())
+        if url_parts.scheme != self.SCHEME:
+            raise ValueError(f"Invalid URI scheme, expected {self.SCHEME}")
+
+        if self.NETLOC not in url_parts.netloc:
+            raise ValueError(f"Invalid URI netloc, expected {self.NETLOC}")
+
+        self._uri_path = url_parts.path
+
+        self.store_id, self.resource_type, self.resource_id, self.file_name, *_ =\
+            self._uri_path.strip("/").split("/") + [self.DEFAULT_FILE_NAME, None]
+
+        if self.resource_type != self.IDENTIFIER:
+            raise ValueError(f"Invalid URI path for reference, {self.resource_type} is not {self.IDENTIFIER}")
+
+
+class ReadSetUri(OmicsUri):
```

Comment:
Most Python devs prefer to avoid inheritance if possible.  It seems like we could implement this with a single `OmicsUri` class that has a `FileType` property.  There is already an [OmicsFileType](https://github.com/awslabs/amazon-omics-tools/blob/main/omics/transfer/__init__.py#L69) enum in the Transfer Manager code.

Not sure that you'd want to reference that from the Transfer Manager, but it might make sense to create a package for common structures.

Reply:
Disagree that we should implement with a single ```OmicsUri``` class. Within the ```ReferenceUri``` and ```ReadSetUri``` we have validation functions which which check if the URI passed in is a valid reference/readset URI. Since each validation function is slightly different, we should put separate out the filestypes as different classes instead of using an enum field

Reply:
Consider that these new classes represent the public interface to Omics Tools, and cannot be refactored.  In general, smaller and simpler interfaces are easier to update in the future as new features are added to Omics.

In this case, it seems that these 4 classes could be replaced by a single `OmicsUri` class that takes the URI as a parameter on the constructor.

Reply:
Okay - I've refactored the code to include all the URI types in one classs

Reply:
Thanks.